core/runtime/impl/block_builder_impl.cpp

```diff
@@ -0,0 +1,138 @@
+  }
+
+  outcome::result<primitives::BlockHeader> BlockBuilderImpl::finalize_block() {
+    OUTCOME_TRY(res,
+                executor_.call(state_code_, "BlockBuilder_finalize_block", {}));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
+
+    WasmMemoryStream s(memory_);
+    OUTCOME_TRY(s.advance(res_addr));
+
+    OUTCOME_TRY(header, codec_->decodeBlockHeader(s));
+
+    return std::move(header);  // warning from clang-tidy without move
+  }
+
+  outcome::result<std::vector<primitives::Extrinsic>>
+  BlockBuilderImpl::inherent_extrinsics(primitives::InherentData data) {
+    OUTCOME_TRY(enc_data, codec_->encodeInherentData(data));
+
+    runtime::SizeType data_size = enc_data.size();
+    runtime::WasmPointer ptr = memory_->allocate(data_size);
+    memory_->storeBuffer(ptr, enc_data);
+
+    wasm::LiteralList ll{Literal(ptr), Literal(data_size)};
+
+    OUTCOME_TRY(
+        res,
+        executor_.call(state_code_, "BlockBuilder_inherent_extrinsics", ll));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
+
+    WasmMemoryStream s(memory_);
+    OUTCOME_TRY(s.advance(res_addr));
+
+    using std::placeholders::_1;
+    auto decode_f = [this](common::ByteStream &stream) {
+      return codec_->decodeExtrinsic(stream);
+    };
+    OUTCOME_TRY(decoded_res, decodeCollection<Extrinsic>(s, decode_f));
+
+    return std::move(decoded_res);
+  }
+
+  outcome::result<CheckInherentsResult> BlockBuilderImpl::check_inherents(
+      primitives::Block block, primitives::InherentData data) {
+    OUTCOME_TRY(enc_block, codec_->encodeBlock(block));
+    OUTCOME_TRY(enc_data, codec_->encodeInherentData(data));
+
+    runtime::SizeType block_size = enc_block.size();
+    runtime::WasmPointer block_ptr = memory_->allocate(block_size);
+    memory_->storeBuffer(block_ptr, enc_block);
+
+    runtime::SizeType data_size = enc_data.size();
+    runtime::WasmPointer data_ptr = memory_->allocate(data_size);
+    memory_->storeBuffer(data_ptr, enc_data);
+
+    wasm::LiteralList ll{Literal(block_ptr), Literal(data_ptr)};
+
+    OUTCOME_TRY(
+        res, executor_.call(state_code_, "BlockBuilder_check_inherents", ll));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
+
+    WasmMemoryStream s(memory_);
+    OUTCOME_TRY(s.advance(res_addr));
+
+    scale::TypeDecoder<bool> decoder;
+    OUTCOME_TRY(ok, decoder.decode(s));
+    OUTCOME_TRY(is_fatal, decoder.decode(s));
+    OUTCOME_TRY(error_data, codec_->decodeInherentData(s));
+
+    return CheckInherentsResult{ok, is_fatal, error_data};
+  }
+
+  outcome::result<common::Hash256> BlockBuilderImpl::random_seed() {
+    OUTCOME_TRY(res,
+                executor_.call(state_code_, "BlockBuilder_random_seed", {}));
+
+    uint32_t res_addr = static_cast<uint64_t>(res.geti64()) >> 32ul;
+
+    WasmMemoryStream s(memory_);
+    OUTCOME_TRY(s.advance(res_addr));
+
+    common::Hash256 hash;
+    for (size_t i = 0; i < common::Hash256::size(); i++) {
```

Comment:
Temporary decode hash manually, because TypeDecoder cannot be used here

Reply:
It's not seen outside scale_codec_impl unit.

Reply:
Though perhaps extern template would work. Anyway, after scale codec is refactored, this place needs to be changed